RCHAIN-3365

tymm opened the pull request:
## Overview
<sup>_What this PR does, and why it's needed_</sup>



### JIRA ticket:
<sup>_Create it if there isn't one already._</sup>



### Notes
<sup>_Optional. Add any notes on caveats, approaches you tried that didn't work, or anything else._</sup>



### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

Commit:
Got Timm's code compiling

Commit:
Forgot to add this file.

Commit:
Make code pass PoSSpec

Commit:
Handle invalid block sender in deposit contract of PoS

Commit:
Swap forEach for parMap

tymm committed:
Make tests compile

tymm committed:
Increase timeout for PoSSpec

tymm committed:
Adjust `refundDeploy` to make use of per-validator vaults

tymm committed:
Use valid validator for block data

tymm committed:
Remove all `Sender is not an active validator` errors

tymm commented:
bors try

zsluedem commented on casper/src/main/resources/PoS.rhox:
```diff
@@ -246,21 +315,26 @@ new  PoS,
                       @state, mvarResultCh <- mvarProcessCh) {
+                    // Before choosing new active validators and paying out any withdrawers,
+                    // deposits have to be accumulated into one vault for now.
+                    accumulateDeposits!(posRevAddress, *accDepositsDoneCh) |
```
I want to state the logic here to make sure I understand correctly. 
Transfer all the all the bonding deposit to a `posRevAddress` and when a validator is unbonding the node, the `posRevAddress` put back the stake

tymm replied:
Yes, that sounds correct.

zsluedem commented on casper/src/main/scala/coop/rchain/casper/genesis/contracts/ProofOfStake.scala:
```diff
@@ -14,7 +14,8 @@ final case class ProofOfStake(
     epochLength: Int,
-    quarantineLength: Int
+    quarantineLength: Int,
+    numberOfActiveValidators: Int = 100
```
It is better to put the `numberOfActiveValidators` in th configuration.

tymm replied:
Which configuration?

zsluedem reviewed (commented):
Good work although the test failed. Leave some questions here.

Commit:
Refactor BlockData handling

Commit:
Make tests that depend on BlockData timestamp pass

Commit:
Fixes 'Make tests that depend on BlockData timestamp pass' - turns out these tests also depend on seqNum

Commit:
Make `createBlock` take a valid `BlockData` (harder to make a mistake)

marcin-rzeznicki commented on rholang/src/main/scala/coop/rchain/rholang/interpreter/SystemProcesses.scala:
```diff
@@ -245,10 +245,17 @@ object SystemProcesses {
           for {
             data <- blockData.get
-            _    <- produce(Seq(RhoType.Number(data.blockNumber), RhoType.Number(data.timeStamp)), ack)
+            _ <- produce(
```
My only issue here is: would it be beneficial to make it a `SystemDeploy`? This would start to move us towards having only single mechanism of Scala <-> RhoLang interop (now we have like 3: `ConstructDeploy`, `SystemProcess` and `SystemDeploy`)

tgrospic replied:
@marcin-rzeznicki do you mean to define `SystemProcess`es part of `SystemDeploy`? How this will work with non-blockchain Rholang?

tymm replied:
@marcin-rzeznicki system processes are called from Rholang. How would you call a system deploy from Rholang?

marcin-rzeznicki replied:
I'm not sure if I'm understanding it correctly, so correct me if I'm wrong. But, at least in this case, the system process simply binds `Ref` to a channel.  I think it can be perfectly simulated in RhoLang. I see similar (?) examples in PoS eg. `state.set("allBonds",state.get("allBonds").set(userPk,amount))` so I'm concluding it can be made into a SystemDeploy

tymm replied:
Now I see what you mean. Since we can't move all system processes to system deploys, I'm not sure if it is worth a refactoring.

marcin-rzeznicki replied:
Ok let's leave it for now

tymm commented:
bors try

tymm commented on casper/src/test/scala/coop/rchain/casper/util/rholang/InterpreterUtilTest.scala:
```diff
@@ -30,40 +29,53 @@ import monix.eval.Task
+              dag,
+              runtimeManager,
+              BlockData(
+                now,
```
This does the job for now but does not reflect the reality where timestamps would increase from block to block. On the other hand smart contracts can't rely on the correctness of timestamp data anyway so having a constant time in tests might be a good thing.

tymm commented on casper/src/test/scala/coop/rchain/casper/util/rholang/InterpreterUtilTest.scala:
```diff
@@ -667,11 +674,13 @@ class InterpreterUtilTest
           dag2           <- blockDagStorage.getRepresentation
           validateResult <- validateBlockCheckpoint[Task](block, dag2, runtimeManager)
           Right(tsHash)  = validateResult
+          _              <- timeEff.advance()
```
Nice!

tymm commented on casper/src/test/scala/coop/rchain/casper/util/rholang/InterpreterUtilTest.scala:
```diff
@@ -69,7 +73,7 @@ class InterpreterUtilTest
               BlockData(
                 now,
-                0,
+                seqNum,
```
I'm unsure if it is okay to equate sequence numbers with block numbers.
@itegulov can you verify that this is okay?

tgrospic replied:
More general question: how to get last block number required for deploy as `validAfterBlockNumber`? https://github.com/rchain/rchain/pull/2811#pullrequestreview-323624096

marcin-rzeznicki replied:
I think it's OK for tests because of: 
```scala
      nextId           = if (block.seqNum == 0) currentId + 1L else block.seqNum.toLong
      newPostState     = block.body.state.copy(blockNumber = nextId)
```
in the `IndexedBlockDagStorage` (thanks for pointing out to it @tymm )

Commit:
itests: Skip java version check in launcher script

The reason is that by default, the script generated by
sbt-native-packager swallows stderr when checking for java version[1],
because for some reason "java -version" prints to stderr. So when java
fails for whatever reason it only prints "No java installation was
detected".

[1] https://github.com/sbt/sbt-native-packager/blob/37d55355/src/main/resources/com/typesafe/sbt/packager/archetypes/scripts/bash-template#L275

Commit:
itests: Rename bootstrap startup functions

Since started_bootstrap() only waits for P2P server to be ready
("Listening on ..."), the other start function names were confusing.

  * Old: ready_bootstrap
    New: started_bootstrap_with_network

  * Old: docker_network_with_started_bootstrap
    New: ready_bootstrap_with_network

Also replace docker_network() followed by started_bootstrap() with
started_bootstrap_with_network().

Also replace (in new names) started_bootstrap_with_network() with
ready_bootstrap_with_network() where appropriate so that the test waits
for bootstrap's approval block creation ('Making a transition to Running
state') instead of just startup of P2P server ('Listening on').

When a peer connects to a bootstrap that hasn't yet created approved
block, it'll enter eternal sleep. Example log of such node:

    Picked up _JAVA_OPTIONS: -Xms2G -Xmx4G
    14:38:56.181 [main] INFO  coop.rchain.node.Main$ - RChain Node 0.9.17-27-g28ede8d (28ede8d8a53b1891c7c0d21b9bb25330fdedc7c9)
    14:38:56.191 [main] INFO  coop.rchain.node.Main$ - Starting with profile docker
    14:38:56.191 [main] INFO  coop.rchain.node.Main$ - Using configuration file: /var/lib/rnode/rnode.conf
    14:38:56.191 [main] WARN  coop.rchain.node.Main$ - Configuration file not found!
    14:38:56.192 [main] INFO  coop.rchain.node.Main$ - Running on network: testnet
    14:38:56.212 [main] INFO  coop.rchain.node.NodeEnvironment$ - Using data dir: /var/lib/rnode
    14:38:56.242 [main] INFO  c.r.c.t.GenerateCertificateIfAbsent - No certificate found at path /var/lib/rnode/node.certificate.pem
    14:38:56.245 [main] INFO  c.r.c.t.GenerateCertificateIfAbsent - Generating a X.509 certificate for the node
    14:38:56.248 [main] INFO  c.r.c.t.GenerateCertificateIfAbsent - Generating a PEM secret key for the node
    14:38:57.974 [main] WARN  c.r.b.dag.PersistentIndex$ - CRC file /var/lib/rnode/dagstorage/latestMessagesCrcPath did not contain a valid CRC value
    14:38:58.025 [main] WARN  c.r.b.dag.PersistentIndex$ - CRC file /var/lib/rnode/dagstorage/blockMetadataCrcPath did not contain a valid CRC value
    14:38:58.065 [main] WARN  c.r.b.dag.PersistentIndex$ - CRC file /var/lib/rnode/dagstorage/equivocationsTrackerCrcPath did not contain a valid CRC value
    14:38:58.093 [main] WARN  c.r.b.dag.PersistentIndex$ - CRC file /var/lib/rnode/dagstorage/invalidBlocksCrcPath did not contain a valid CRC value
    14:38:58.115 [main] WARN  c.r.b.dag.PersistentIndex$ - CRC file /var/lib/rnode/dagstorage/blockHashesByDeployCrcPath did not contain a valid CRC value
    14:39:02.214 [node-runner-33] INFO  c.r.c.engine.CasperLaunch$$anon$1 - ApprovedBlock not found in storage, connecting to existing network
    14:39:02.232 [node-runner-33] INFO  coop.rchain.node.NodeRuntime - Starting node that will bootstrap from rnode://26088a91317d0dc77a909eddcc353fca91906367@rchain-symmx.bootstrap?protocol=40400&discovery=40404
    14:39:02.614 [node-runner-33] INFO  coop.rchain.node.NodeRuntime - External API server started at rchain-symmx.bonded-validator:40401
    14:39:02.619 [node-runner-33] INFO  coop.rchain.node.NodeRuntime - Internal API server started at rchain-symmx.bonded-validator:40402
    14:39:02.625 [node-runner-33] INFO  coop.rchain.node.NodeRuntime - Kademlia RPC server started at rchain-symmx.bonded-validator:40404
    14:39:03.318 [node-runner-33] INFO  coop.rchain.node.NodeRuntime - Listening for traffic on rnode://d5815467e0426d9ac8703496ca81a3d984aebbc1@rchain-symmx.bonded-validator?protocol=40400&discovery=40404.
    14:39:03.346 [node-runner-33] INFO  coop.rchain.node.NodeRuntime - Waiting for first connection.
    14:39:05.050 [tl-dispatcher-server-queue-89] INFO  coop.rchain.comm.rp.Connect$ - Peers: 1
    14:39:05.070 [tl-dispatcher-server-queue-115] INFO  coop.rchain.comm.rp.HandleMessages$ - Responded to protocol handshake request from rnode://d5815467e0426d9ac8703496ca81a3d984aebbc1@rchain-symmx.bonded-validator?protocol=40400&discovery=40404
    14:39:13.388 [node-runner-43] INFO  c.rchain.casper.util.comm.CommUtil$ - Starting to request ApprovedBlockRequest
    14:39:13.463 [node-runner-33] INFO  c.rchain.casper.util.comm.CommUtil$ - Successfully sent ApprovedBlockRequest to rnode://26088a91317d0dc77a909eddcc353fca91906367@rchain-symmx.bootstrap?protocol=40400&discovery=40404
    14:39:13.483 [tl-dispatcher-server-queue-126] INFO  c.r.comm.transport.StreamObservable - Streaming packet (type = NoApprovedBlockAvailable) to peers rnode://d5815467e0426d9ac8703496ca81a3d984aebbc1@rchain-symmx.bonded-validator?protocol=40400&discovery=40404
    14:39:13.721 [tl-client-stream-queue-103] INFO  c.r.c.transport.GrpcTransportClient - Streamed packet /var/lib/rnode/tmp/comm/20191204143913_cfcea60c_packet.bts to rnode://d5815467e0426d9ac8703496ca81a3d984aebbc1@rchain-symmx.bonded-validator?protocol=40400&discovery=40404
    14:39:13.756 [grpc-io-79] INFO  coop.rchain.casper.engine.Engine$ - No approved block available on node rnode://26088a91317d0dc77a909eddcc353fca91906367@rchain-symmx.bootstrap?protocol=40400&discovery=40404

Without waiting for bootstrap's approved block before starting peers,
there is a race condition. Because of it integration tests sometimes
take multiple of full start timeout (30 minutes).

tymm commented:
bors r+

tymm committed:
Increase timeout for PoSSpec once again

tymm commented:
bors r+

tymm committed:
Change wait_for_approved_block to true

tymm commented:
bors r+

zsluedem commented:
bors try

Commit:
itests: test_heterogenous_validators -> wait for bootstrap's approved block

tymm commented:
bors r+

tymm committed:
Skip invalid integration test

tymm commented:
bors r+

tymm commented:
bors retry

tymm commented:
bors try-

tymm commented:
bors r-

tymm commented:
bors r+
